app/code/Magento/Catalog/Model/Product/Copier.php

```diff
@@ -141,20 +164,16 @@ private function setStoresUrl(Product $product, Product $duplicate) : void
             if ($useDefault) {
                 continue;
             }
-            $isDuplicateSaved = false;
+
             $duplicate->setStoreId($storeId);
             $urlKey = $productResource->getAttributeRawValue($productId, 'url_key', $storeId);
+
             do {
                 $urlKey = $this->modifyUrl($urlKey);
                 $duplicate->setUrlKey($urlKey);
-                $duplicate->setData('url_path', null);
-                try {
-                    $duplicate->save();
-                    $isDuplicateSaved = true;
-                    // phpcs:ignore Magento2.CodeAnalysis.EmptyBlock
-                } catch (\Magento\Framework\Exception\AlreadyExistsException $e) {
-                }
-            } while (!$isDuplicateSaved);
+            } while (!$attribute->getEntity()->checkAttributeUniqueValue($attribute, $duplicate));
```

Comment:
This actually not solving an issue with infinite loop.
Could you add stopper, if we tried 10 times - then stop this loop and throw an exception.

Reply:
How it can become infinite? What's the reason to change previous logic?

New logic is not race condition-safe.

Reply:
@orlangur It becomes an infinite loop when following the manual testing scenario. Will add the stopper later this week if that's the preferred solution.

Reply:
> New logic is not race condition-safe.

That's true. Is there any ideas how to fix that?


Related: running `->save` method in a loop also causing duplicating product images in the media directory from original product.

So in combination with infinite loop - we're having issue that disk space utilization grows until process will be killed.

Reply:
@orlangur 
I just analyzed deeper.
In case of race condition - we'll have exception during saving the product and product will not be duplicated anyway. I think this is much better result than infinite loop.